<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="sim" default="false" />
  <arg name="debug" default="false" />

  <!-- Arg to start the datalogger -->
  <arg name="datalogger" default="false" />

  <!-- Args to start different capabilities -->
  <arg name="start_all" default="false" />
  <arg name="navigation" default="$(arg start_all)" />
  <arg name="perception" default="$(arg start_all)" />
  <arg name="manipulation" default="$(arg start_all)" />
  <arg name="sound" default="$(arg start_all)" />
  <arg name="task_executor" default="$(arg start_all)" />
  <arg name="task_monitor" default="$(arg task_executor)" />

  <!-- Common args -->
  <arg name="cloud_topic" value="/head_camera/depth_registered/points" />
  <arg name="grasp_classifier_file" value="random_forest.pkl" />
  <arg name="segmentation_config" value="$(find task_executor)/config/zones.yaml" />
  <arg name="task_config" value="$(find task_executor)/config/tasks.yaml" />

  <!-- Simulation vs. non simulation args -->
  <arg name="data_config" value="$(find task_executor)/config/simulation.yaml" if="$(arg sim)" />
  <!-- <arg name="map_file" value="$(find fetch_gazebo_demo)/maps/test_zone.yaml" if="$(arg sim)" /> -->
  <arg name="data_config" value="$(find task_executor)/config/robot.yaml" unless="$(arg sim)" />
  <!-- <arg name="map_file" value="$(find fetch_demos)/maps/map.yaml" unless="$(arg sim)" /> -->

  <!-- Other args -->

  <!-- Navigation -->
  <group if="$(arg navigation)">
    <!-- <include file="$(find fetch_navigation)/launch/fetch_nav.launch">
      <arg name="map_file" value="$(arg map_file)" />
    </include> -->
  </group>

  <!-- Perception -->
  <group if="$(arg perception)">
    <!-- RAIL Segmentation -->
    <node pkg="rail_segmentation" type="rail_segmentation" name="rail_segmentation" output="screen">
      <param name="zones_config" value="$(arg segmentation_config)" />
      <param name="point_cloud_topic" value="$(arg cloud_topic)" />
      <param name="max_cluster_size" value="30000" />
      <param name="min_cluster_size" value="75" />
      <param name="crop_first" value="true" />
    </node>

    <!-- Grasp Suggestion -->
    <include file="$(find fetch_grasp_suggestion)/launch/grasp_suggestion.launch">
      <arg name="cloud_topic" value="$(arg cloud_topic)" />
      <arg name="classifier_file" value="$(arg grasp_classifier_file)" />
    </include>

    <!-- Fetchit Bin Detection -->
    <include file="$(find fetchit_bin_detector)/launch/launch_detector.launch">
      <arg name="launch_segmentation" value="false" />
      <arg name="pcl_topic" value="$(arg cloud_topic)" />
      <arg name="seg_service" value="/rail_segmentation/segment_objects" />
      <arg name="viz_detections" value="$(arg debug)" />
    </include>
  </group>

  <!-- Manipulation -->
  <group if="$(arg manipulation)">
    <!-- common to all (sampling-based) manipulation -->
    <include file="$(find fetch_moveit_config)/launch/move_group.launch">
      <arg name="info" value="true"/>
      <arg name="allow_active_sensing" value="true" />
    </include>
    <node pkg="manipulation_actions" type="collision_scene_manager" name="collision_scene_manager" output="screen" />
    <!-- Set a static transform for the SCHUNK machine in simulation to make
    MoveIt! happy -->
    <node name="world_to_odom_publisher" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 world odom 100" if="$(arg sim)"/>

    <!-- task-specific manipulation actions -->
    <node pkg="fetch_grasp_suggestion" type="executor" name="grasp_executor" output="screen" />
    <node pkg="manipulation_actions" type="cluttered_grasper" name="cluttered_grasper" output="screen" />
    <node pkg="manipulation_actions" type="in_hand_localizer" name="in_hand_localizer" output="screen" />
    <node pkg="manipulation_actions" type="placer" name="placer" output="screen" />
  </group>

  <!-- Sound Play -->
  <group if="$(arg sound)">
    <node name="sound_server" pkg="sound_interface" type="server_node.py" />
  </group>

  <!-- Data Logger -->
  <node name="datalogger" pkg="task_monitor" type="datalogger_node.py" output="screen" if="$(arg datalogger)" />

  <!-- Task Monitoring and Management -->
  <include file="$(find task_monitor)/launch/monitor.launch" if="$(arg task_monitor)">
    <arg name="sim" value="$(arg sim)" />
  </include>

  <!-- Task Execution -->
  <group if="$(arg task_executor)">
    <include file="$(find task_executor)/launch/task_executor.launch">
      <arg name="sim" value="$(arg sim)" />
      <arg name="task_config" value="$(arg task_config)" />
      <arg name="data_config" value="$(arg data_config)" />
    </include>
  </group>
</launch>
